app/controllers/facility_user_reservations_controller.rb

```diff
@@ -18,35 +19,39 @@ def index
                      .paginate(page: params[:page])
   end
 
-  # PUT /facilities/:facility_id/users/:user_id/reservations/:id/cancel
+  # PUT /facilities/:facility_id/users/:user_id/reservations/:order_detail_id/cancel
   def cancel
-    order_detail = user_order_details.find(params[:id])
-    raise ActiveRecord::RecordNotFound unless order_detail.reservation.try(:can_cancel?)
-    order_detail.transaction do
-      if cancel_with_fee(order_detail)
-        flash[:notice] = text("cancel.success")
-      else
-        flash[:error] = text("cancel.error")
-        raise ActiveRecord::Rollback
-      end
+    reservation = @order_detail.reservation
```

Comment:
I refactored, though I don't know if it's as clear as it could be.

The logic I'm going for is:

* If the `OrderDetail` has no `Reservation`, 404
* If the `Reservation` was already canceled, show the success notice and redirect
* If the `Reservation` is not cancelable, 404
* Attempt to cancel the `Reservation`
  - If cancellation succeeds, show the success notice and redirect
  - If cancellation fails, show the failure notice and roll back